dev/src/bulk-writer.ts

```diff
@@ -668,68 +774,73 @@ export class BulkWriter {
         return lastBatch;
       }
     }
-    return this.createNewBatch();
+
+    return this.createNewBatch(batchQueue);
   }
 
   /**
-   * Creates a new batch and adds it to the BatchQueue. If there is already a
-   * batch enqueued, sends the batch after a new one is created.
+   * Creates a new batch and adds it to the appropriate batch queue. If there
+   * is already a batch enqueued, sends the batch after a new one is created.
    *
    * @private
    */
-  private createNewBatch(): BulkCommitBatch {
+  private createNewBatch(batchQueue: BulkCommitBatch[]): BulkCommitBatch {
     const newBatch = new BulkCommitBatch(
       this.firestore,
       this.firestore.batch(),
-      this.maxBatchSize
+      this._maxBatchSize
     );
 
-    if (this.batchQueue.length > 0) {
-      this.batchQueue[this.batchQueue.length - 1].markReadyToSend();
-      this.sendReadyBatches();
+    if (batchQueue.length > 0) {
+      batchQueue[batchQueue.length - 1].markReadyToSend();
+      this.sendReadyBatches(batchQueue);
     }
-    this.batchQueue.push(newBatch);
+
+    batchQueue.push(newBatch);
     return newBatch;
   }
 
   /**
-   * Attempts to send batches starting from the front of the BatchQueue until a
-   * batch cannot be sent.
+   * Attempts to send batches starting from the front of the provided batch
+   * queue until a batch cannot be sent.
    *
    * After a batch is complete, try sending batches again.
    *
    * @private
    */
-  private sendReadyBatches(): void {
-    const unsentBatches = this.batchQueue.filter(
-      batch => batch.state === BatchState.READY_TO_SEND
-    );
-
+  private sendReadyBatches(batchQueue = this._batchQueue): void {
     let index = 0;
     while (
-      index < unsentBatches.length &&
-      unsentBatches[index].state === BatchState.READY_TO_SEND
+      index < batchQueue.length &&
+      batchQueue[index].state === BatchState.READY_TO_SEND
     ) {
-      const batch = unsentBatches[index];
+      const batch = batchQueue[index];
+
+      // Deferred promise that resolves when the current batch completes.
+      const batchCompletedDeferred = new Deferred<void>();
+      this._pendingBatches.add(batchCompletedDeferred.promise);
 
       // Send the batch if it is under the rate limit, or schedule another
       // attempt after the appropriate timeout.
-      const delayMs = this.rateLimiter.getNextRequestDelayMs(batch.opCount);
+      const delayMs = this._rateLimiter.getNextRequestDelayMs(batch.opCount);
       assert(delayMs !== -1, 'Batch size should be under capacity');
       if (delayMs === 0) {
-        this.sendBatch(batch);
+        this.sendBatch(batch, batchQueue, batchCompletedDeferred);
       } else {
-        delayExecution(() => this.sendReadyBatches(), delayMs);
+        delayExecution(() => {
+          batchCompletedDeferred.resolve();
+          this._pendingBatches.delete(batchCompletedDeferred.promise);
+          this.sendReadyBatches();
```

Comment:
I forgot to mention this last time - I think this would be more future proof if `batchCompleteDeferred` was the last statement in this block.